src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -299,8 +303,37 @@ unsigned FindNextDeviceTypeIndex(Span<const EmberAfDeviceType> types, const Data
 
 const ConcreteCommandPath kInvalidCommandPath(kInvalidEndpointId, kInvalidClusterId, kInvalidCommandId);
 
+// Ember requires a persistence provider
+DefaultAttributePersistenceProvider gDefaultAttributePersistenceProvider;
+
 } // namespace
 
+CHIP_ERROR CodegenDataModelProvider::Startup(DataModel::InteractionModelContext context)
+{
+    ReturnErrorOnFailure(DataModel::Provider::Startup(context));
+
+    // Ember/codegen requires a persistence provider. Ensure one is set.
+    if (GetAttributePersistenceProvider() == nullptr)
+    {
+        ReturnErrorOnFailure(gDefaultAttributePersistenceProvider.Init(GetSafeAttributePersistenceProvider()));
+        SetAttributePersistenceProvider(&gDefaultAttributePersistenceProvider);
```

Comment:
Why is this being handled here?  This seems very much unrelated to just injecting the data model provider...

Reply:
It is a sideffect of initialization order: data model provider initializes the ember persistence, so I need a re-ordering of when each of the persistence providers are initialized/updated.

I found splitting somewhat harder since this is directly required by me passing in the codegendatamodel to the server init instead of having it default-pulled whenever `InteractionModelEngine::GetDataModeProvider` is called.

I can try to split it out, however would prefer not to due to it already being reasonably tangled.

Reply:
I guess here's my confusion: before this PR, there is no mention of attribute persistence in the codegen data model provider at all.  Is this a hidden dependency that is covered up by accidents of initialization order?  But if the codegen data model provider does not use attribute persistence, why should it be its job to initialize it at all?

Reply:
I am trying to get a way for applications to not know about ember at all. One of the final places that I found was that `AttributePersistenceProvider` is ember-specific because it uses `EmberAfAttributeMetadata`. So then a `DefaultAttributePersistenceProvider` ties things to ember and the server having such a member in https://github.com/project-chip/connectedhomeip/blob/master/src/app/server/Server.h#L687 make it so that Server is hard-coding ember usage.

My delta looks to completely separate ember-specific parts in `util` and keep non-ember (but generally usable by anyone) parts still in app - this is the `SafeAttributePersistenceProvider`

Reply:
As discussed on Slack:

This PR is conflating two things:

1. The DataModel::Provider should be injectable.
2. The codegen provider calls some functions that for their proper functioning rely on  `SafeAttributePersistenceProvider` having been called.  But they also rely on `emberAfInit` being called and `emberAfEndpointConfigure` being called, at the very least.

If we want to do a PR that makes sure to move all the "init the Ember stack" bits into one place, great.  But that has nothing to do with whether the DataModel::Provider is injectable, and the injectability can be done either before or after the "move to one place" changes.